pkg/profileparser/profileparser.go

```diff
@@ -658,7 +658,12 @@ func ParseRulesAndDo(contentDom *xmlquery.Node, stdParser *referenceParser, pb *
 				}
 			}
 			if warnings != nil {
-				p.Warning = strings.Join(warnings, "\n")
+				p.Warning, valueRendered, err = utils.RenderValues(utils.XmlNodeAsMarkdownPreRender(rationale), valuesList)
```

Comment:
@Vincent056 sorry but could you tell me where is the best position to add the test for this new feature? `profileparser_test.go` must be the target file, but in which part do I need to add the test?

Reply:
@tmishina maybe create something like this one https://github.com/openshift/compliance-operator/blob/10f5c89c801a5f10b75068ac2da8d9b1b29120cb/pkg/profileparser/profileparser_test.go#L607-L633

Reply:
@Vincent056 I wrote test and the test failed. I check the reason and found this.

https://github.com/openshift/compliance-operator/blob/bb69581e97be73e78bb25dad2ff797de763eea70/pkg/utils/parse_arf_result.go#L517-L521

This code means: when a `xccdf-1.2:warning` contains an element which class is `ocp-api-endpoint`,  such a `warning` element will not be parsed, and therefore substitution function will never be called.

My conclusion here is that we don't need to add a functionality to  to substitute variables in `warning` element at a time of profile parsing. Is my understanding correct?

Reply:
@JAORMX, Do you have any knowledge on this one? If we are using custom variables in warning, I think we should also parse it?

Reply:
> I found important issue at the test with my modified ComplianceAsCode/content.
> 
> The issue is that customized variables specified in a tailored profile resource does not affect the API resource collector, and it collects resources specified as default value for each variable. It ignores customized values in the tailored profile resource. That is not an expected behaviour.
> 
> The reason is that API resource collector checks `warnings` elements in the data stream stored in a ComplianceAsCode/content image (no substitution), not in a profile resource (substituted). (I've not checked whether this also occurs in description/rationale elements)
> 
> As far as I know, there are two solution candidates.
> 
> 1. use my first code (introduce special macro to embed variable reference in a data stream image and parse it in CO)
>    
>    * pros: [small modification](https://github.com/tmishina/compliance-operator/commit/505ebf9de27fe80d71fe7176e62f381fccb56f5a)
>    * cons: substitution mechanism differ than that for description/rationale of `profile` resource
> 2. modify the API resource collector to use `warnings` in kube resource instead of data stream
>    
>    * pros: same substitution mechanism as description/rationale
>      
>      * cons: a number of (maybe 10-100 order) lines should be changed (mainly at [scap.go](https://github.com/openshift/compliance-operator/blob/0a7246b39ac3bffb33961341a4593bf581661543/cmd/manager/scap.go#L242-L248))
> 
> My preference is 1, but @Vincent056, @JAORMX, I'd appreciate it if I can hear your opinions, thanks.

Hi @tmishina, Yes, in the warning and rational section, it only gets values from the input value list when calling `RenderValues`, but I believe you can first get the tailored profile CM, then get the value list, and feed it to the parsing function. in https://github.com/openshift/compliance-operator/blob/441faa0d3fe6d43a4803297b9615bb74bc8dcca6/cmd/manager/scap.go#L186 you can parse the content here before calling `utils.GetPathFromWarningXML(in)`

Reply:
@Vincent056 Thank you for your suggestion! I made changes.
Major update is done to `scap.go`, especially the change on the method signature of `getPathFromWarningXML`.

https://github.com/openshift/compliance-operator/pull/750/files#diff-f86e02e20295a205521712d1773d904635f825c76fff3e0408ad4decc24c578aL191

In addition, I implemented test for the new functionality (overriding variables with the ones in tailored profile).

https://github.com/openshift/compliance-operator/pull/750/files#diff-2cd82c26c0af69487c7dfc803d1b54c1091dc696c3071969c6f3d3d9cc0acf94R108-R139

Are these changes OK?